### PR TITLE
No-allocating queue for MemoryPoolBlock in MemoryPool

### PR DESCRIPTION
**Pull request by Scooletz**
This `PR` provides a custom no-allocating queue for `MemoryPoolBlock` in the `MemoryPool`. 

It works in a bit different way from the `ConcurrentQueue<T>`. The `ConcurrentQueue<T>` uses segments of length `32` and status fields to mark an item as published. These segments requires allocations though.

This PR adds a `Poolable` class that has a private field representing an identifier of the next block, which enables creation of a concurrent store from the blocks themselves. To address ABA problem, instead of blocks references, their `int` identifiers are used, which combined with `int` `abaCounter` can be exchange in one operation using `Interlocked.CompareExchange( ref long, long, long)`.

This PR adds a benchmark for performance and memory allocations. 

| Method | Median | StdDev | Scaled | Scaled-SD | Gen 0 | Gen 1 | Gen 2 | Bytes Allocated/Op |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| PR | 51.4403 ns | 2.8007 ns | 1.00 | 0.07 | - | - | - | 0.00 |
| ConcurrentQueue | 51.6737 ns | 3.1264 ns | 1.00 | 0.00 | 141.63 | 60.70 | 20.23 | 5.45 |

### CODE REVIEW

**Comment by Scooletz** [comment]
@davidfowl Is there any benchmarking test suite I could use to write a custom test for the `MemoryPool`? I'd like to be able to prove the better performance instead of talking it through ;-)

**Comment by Scooletz** [comment]
It looks like the build has hung. The reason might be the number of allocated blocks which is quite high: `4096`. I know it's a lot but it's the easiest way to test the concurrent structure.

**Comment by benaadams** [comment]
Sure it won't suffering from the [ABA problem](https://en.wikipedia.org/wiki/ABA_problem) on the CAS testing? (Which is one of the reasons the ConcurrentQueue uses segments which it discards)

**Comment by Scooletz** [comment]
It suffers from it. Segments might be dropped when the head reappeared as the head with additional values. I'm not closing a PR though, let me try to polish it.

**Comment by Scooletz** [comment]
The current approach will not suffer from ABA @benaadams. I've updated the description to match implementation  . Still, it requires performance tests, so any guidance (is there already a perf suite in channels) would be greatly appreciated.

**Comment by davidfowl** [comment]
@Scooletz There's a new performance test project you can use

File: test/Channels.Tests.Performance/MemoryPoolBenchmark.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                    With(Jit.RyuJit).
+                    With(Runtime.Clr).
+                    WithIterationTime(200). // 200ms per iteration
+                    WithWarmupCount(5).
+                    WithTargetCount(10));
+                Add(new MemoryDiagnoser());
+            }
+        }
+
+        public const int InnerLoopCount = 10000;
```
Comment (benaadams) [inline test/Channels.Tests.Performance/MemoryPoolBenchmark.cs]: Looking at `Leases` count is 4? `OperationsPerInvoke` divides the total time by its value for when you add multiples of the operation in the function.

Comment (Scooletz) [reply test/Channels.Tests.Performance/MemoryPoolBenchmark.cs]: Thx @benaadams Will update my test and rerun.

**Comment by benaadams** [comment]
Looking a bit closer, rather than changing `MemoryPool` you might want to add a new type that either inherits from `MemoryPool` or implements `IBufferPool`

Then in the perf tests you can compare `MemoryPool` to this version; the channels themselves [take an `IBufferPool`](https://github.com/davidfowl/Channels/blob/master/src/Channels/ChannelFactory.cs#L16-L23) so side-by-side implementations of the pools can coexist.

**Comment by Scooletz** [comment]
@benaadams I run the tests with the previous version (with the added benchmark) and then on the updated version. I did 3 runs each, selecting the best result. Let me present the results first to bring some numbers to the table.

#### Previous implementation

 Method |     Median |    StdDev | Gen 0 | Gen 1 | Gen 2 | Bytes Allocated/Op |
------- |----------- |---------- |------ |------ |------ |------------------- |
 Leases | 64.0265 ns | 1.0048 ns | 89.00 | 67.00 | 24.00 |               5.76 |

#### New no-alloc implementation

 Method |     Median |    StdDev | Gen 0 | Gen 1 | Gen 2 | Bytes Allocated/Op |
------- |----------- |---------- |------ |------ |------ |------------------- |
 Leases | 58.3106 ns | 1.0316 ns |     - |     - |  2.00 |               0.11 |

**Comment by Scooletz** [comment]
@benaadams @davidfowl Thoughts?

**Comment by Scooletz** [comment]
@benaadams @davidfowl I separated the `Poolable` class and compared that with a concurrent queue. It looks like there's not that much improvement in it beside removing all the allocations.

**Comment by davidfowl** [comment]
This has moved to corefxlab

**Pull request closed without merging**